### PR TITLE
Prepare PVFs if node is a validator in the next session

### PR DESCRIPTION
**Pull request by AndreiEres**
Closes https://github.com/paritytech/polkadot-sdk/issues/4324
- On every active leaf candidate-validation subsystem checks if the node is the next session authority.
- If it is, it fetches backed candidates and prepares unknown PVFs.
- We limit number of PVFs per block to not overload subsystem.

### CODE REVIEW

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (alexggh) [inline polkadot/node/core/candidate-validation/src/lib.rs:273]: Nit: No need for this clone ctx.sender() &mut already.

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (alexggh) [inline polkadot/node/core/candidate-validation/src/lib.rs:284]: Doesn't this return `true` if we are authorithy in the current session, if that's the case we need to make sure we don't prepare them then, because we will create a storm of prepares that would incapacitate the node.

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (alexggh) [inline polkadot/node/core/candidate-validation/src/lib.rs:279]: Why clearing here? Once we prepare it should be there until restart.

**Review by alexggh — COMMENTED** [review]
First pass overall looks good, left you some comments.

On top of that we need some tests as well that test the following invariants:
1. The logic doesn't get triggered if we are in active set.
2. Checks that when we enter that active set no compilation is triggered because we already compiled them.

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (sandreim) [inline polkadot/node/core/candidate-validation/src/lib.rs:366]: Yes, we are only interested in future.

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (sandreim) [reply polkadot/node/core/candidate-validation/src/lib.rs:284]: This needs to be changed to account for the fact we might be an authority in the current session which is what tipically happens in practice. 

And yeah, we could end up preparing more artifacts than needed and we need to fix that.

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (sandreim) [inline polkadot/node/core/candidate-validation/src/lib.rs:462]: Considering `HOST_MESSAGE_QUEUE_SIZE` is 10 can this block ?

Comment (sandreim) [reply polkadot/node/core/candidate-validation/src/lib.rs:462]: It could be a problem if we stall the main loop here while we are active validator. Processing of any incoming validation requests will also be delayed by up to 100 (n_cores) validation code decompressions.

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (s0me0ne-unkn0wn) [reply polkadot/node/core/candidate-validation/src/lib.rs:279]: I thought the same, but chances are that by this point, the artifact might have been pruned because it had been unused for more than 24 hours. So clearing here is more foolproof than not clearing, but some better solution may also be considered.

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (s0me0ne-unkn0wn) [inline polkadot/node/core/candidate-validation/src/lib.rs:423]:
```suggestion
	let timeout = pvf_prep_timeout(&executor_params, PvfPrepKind::Prepare);
```

We're preparing for execution, so the shorter timeout must be used.

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (s0me0ne-unkn0wn) [inline polkadot/node/core/candidate-validation/src/lib.rs:555]: That's a good question indeed: What set of executor parameters should be used to compile the PVFs in advance? There may already be something new in `pending_config` that will become effective in the next session, and all this in-advance compilation will be in vain in that case. That's a corner case, as changes to the executor parameters are extremely rare, so I do not insist it should be addressed in this very PR, but it's something that might probably be considered in the future. Generally, it's not the first time we're having hard times trying to predict next session's executor parameters: #694

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (AndreiEres) [reply polkadot/node/core/candidate-validation/src/lib.rs:277]: I supposed we should start from scratch on every new session. It’s just a mechanism to not overload the pvf host with unnecessary checking.

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (AndreiEres) [reply polkadot/node/core/candidate-validation/src/lib.rs:553]: It might not be so bad to precompile pvf anyway. Anyway, the validator will have to compile a new one in the next session. Or we can consider a smarter solution?

**Review by alexggh — APPROVED** [review]
Looks good to me, I would also test with an integration test, to make sure the changes have the desired effect.

File: polkadot/zombienet_tests/smoke/0005-precompile-pvf-smoke.js
Comment (pepoviola) [inline polkadot/zombienet_tests/smoke/0005-precompile-pvf-smoke.js:1]: This file isn't used as part of the test. 
```
alice: js-script ./0003-deregister-register-validator.js with "register,dave" return is 0 within 30 secs
```

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (s0me0ne-unkn0wn) [reply polkadot/node/core/candidate-validation/src/lib.rs:632]: I'm not 100% sure I'm not missing something, but it seems like we could use `pending_config` from the `configuration` pallet to get the executor environment params. That gives us two benefits:
1. We're always getting params that will be effective in the future session;
2. We get it directly from storage, thus saving a couple of runtime API calls.

@sandreim WDYT?

Comment (sandreim) [reply polkadot/node/core/candidate-validation/src/lib.rs:632]: Yes, we already know that in session N we are gonna become a parachain validator. So, we need to prepare the artifacts using the same execution params from session N.

I think 1 is preferable as a proper interaface, rather than reading storage directly.

Alternatively, we could could introduce a new runtime API that returns the execution parameters for a given session. This should solve future problems with getting the next session executor params in one place.

Comment (AndreiEres) [reply polkadot/node/core/candidate-validation/src/lib.rs:632]: @sandreim @s0me0ne-unkn0wn I think it can be a good job for an additional PR. What do you think if I move this conversation to an issue and start doing that after we merged this one?

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (s0me0ne-unkn0wn) [reply polkadot/node/core/candidate-validation/src/lib.rs:532]: ARAIR the whole subsystem has been moved to the blocking pool (#3122)
Yes, it would indeed be good to offload those decompressions to a separate task on a blocking pool; we discussed that. At that point, it seemed like firing cannons at sparrows. But in the context of #5012 it makes much more sense now.
Still, that's out of the scope of this issue. Raised #5071.

Comment (AndreiEres) [reply polkadot/node/core/candidate-validation/src/lib.rs:532]: Returning to the original question. We don't do any work when we're an active validator. In other cases we allow to prepare only one PVF per block to not load the validator. It's plenty of time to prepare all PVFs during the session if a validator connected in the beginning but don't overload if it connected in the end.

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (s0me0ne-unkn0wn) [reply polkadot/node/core/candidate-validation/src/lib.rs:632]: After recalling how it works, I agree. That will require changing the runtime API, putting that into the staging API, then bumping its version, etc. Not an easy change. Raised #5080.

Comment (AndreiEres) [reply polkadot/node/core/candidate-validation/src/lib.rs:632]: After a conversation with @sandreim:
- This concern is valid and not only for just connected validators but for all. See https://github.com/paritytech/polkadot-sdk/issues/4203
- It's OK to solve this issue separately. 

@s0me0ne-unkn0wn what do you think?

Comment (sandreim) [reply polkadot/node/core/candidate-validation/src/lib.rs:632]: I'm ok with implementing https://github.com/paritytech/polkadot-sdk/issues/4203 in a separate PR which applies to these changes as well.  

Getting this one in the current release if we can still make it to backport there would be great. I'd want to wait for validators to upgrade to this fix before bumping Polkadot validator count to 400.

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (sandreim) [inline polkadot/node/core/candidate-validation/src/lib.rs:398]: Do we still want to do this now ?

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (sandreim) [inline polkadot/node/core/candidate-validation/src/lib.rs:508]: I think this just keeps decompressing needlessly if we are a validator intermittently: at session 1, then at session 2 we are no longer, but we are again at session 3. 

All that was prepared for session 1 will be reset as soon as we go into session 2 (`already_prepared_code_hashes` is cleared). Is there a better way to check? Like query the PVF subsystem if it has it already ? Also I think the query needs to take into account the executor params which we are preparing with not just code hash.

Comment (sandreim) [reply polkadot/node/core/candidate-validation/src/lib.rs:507]: let's handle it a new PR.

**Review by sandreim — APPROVED** [review]
Thanks @AndreiEres. LGTM modulo the optimisations that were discussed , but you are already on them in other PRs.  

Let's burn this in on our Kusama validators until the release comes out, just to get more data from real world sooner.

**Review by s0me0ne-unkn0wn — APPROVED** [review]
LGTM :+1:

File: polkadot/node/core/candidate-validation/src/lib.rs
Comment (s0me0ne-unkn0wn) [reply polkadot/node/core/candidate-validation/src/lib.rs:507]: > let's handle it a new PR.

Or maybe will not be even needed: https://github.com/paritytech/polkadot-sdk/issues/5071#issuecomment-2241014329

Comment (sandreim) [reply polkadot/node/core/candidate-validation/src/lib.rs:507]: I think we still need to polish it. #5071 ticket talks about the other path, where we actually do candidate validation, here we just try to compile artifacts for blocks backed on chain. Given that, I expect decompression should be fast, the only thing concerning here is that we do it for all blocks.

**Pull request merged**
